### PR TITLE
various credential improvements (including initrd creds, creds in generators, fstab + getty creds)

### LINKED ISSUE #27260: Support defining mount and swap units via credentials
### Component

other

### Is your feature request related to a problem? Please describe

Linux today supports a variety of block and file system snapshot mechanisms (LVM2, btrfs, split mirrors etc.) and it is possible to boot a simple system with just a single (root) file system into a snapshot of the system state using the existing `root=` and `rootflags=` kernel command line options.

Being able to boot into a snapshot has application for rolling back unsuccessful updates, system recovery and troubleshooting.

It is not currently directly possible to boot into a snapshot state for more complex file system layouts (for e.g. with separate `/var`, `/home` etc.) without first modifying the snapshot content to reference the appropriate block devices or subvolumes in `/etc/fstab`.

### Describe the solution you'd like

The ability to specify mount and swap units directly on the kernel command line. This would allow (in combination with `fstab=no`) the overriding of the file system mount structure for a given boot of the operating system image.

This allows a complete snapshot of the system state to be booted for e.g. in the case of a failed upgrade that renders the system unbootable.

The user can then roll back to the snapshot state minimizing down time and providing a valuable safety net around updates to the operating system.

The syntax would be something like:

```
  systemd.mount=what:where:fstype:options

  systemd.swap=what:options
```
Where a literal ':' in any of the fields is escaped as '::'.

This could be implemented as a new generator or by extending the existing fstab-generator.

### Describe alternatives you've considered

It's possible to implement these options in a separate component that provides a new system generator to recognise and act on the unit syntax. This would require users to install additional packages to take advantage of the feature.

It's also possible to mount the snapshot of the root file system and directly modify /etc/fstab however this complicates rollback (since the changes to the fstab must either be reverted, or preserved in a separate block or file system snapshot).

### The systemd version you checked that didn't have the feature you are asking for

253

### LINKED ISSUE #27604: Confidential computing: revise trusting smbios and other hypervisor-driven payloads
In the confidential computing case the hypervisor is untrusted, so we need to establish a way to detect that case, and stop trusting smbios oem strings and unsigned payloads in the ESP by default.

https://github.com/systemd/systemd/pull/27358#issuecomment-1542129454
https://github.com/systemd/systemd/pull/27358#issuecomment-1542144948
https://github.com/systemd/systemd/pull/27358#issuecomment-1542223091

A suggestion on how to do so is to check MSRs, see "15.34.10 SEV_STATUS MSR" in https://www.amd.com/system/files/TechDocs/24593.pdf

In terms of what needs changing, smbios kernel command line extensions, as they are passed directory by the hypervisor, and unsigned extensions/credentials in the ESP are at the top of the list, as the ESP is not encrypted and the hypervisor can drop payloads there, and they are used before ExitBootServices.

For the userspace stuff (eg: credentials), perhaps we could make them disabled-by-default, but allow opt-in? Userspace code and configuration is signed or encrypted, and thus should be fine to trust to make that decision (with a you-keep-the-pieces warning).

### PR DESCRIPTION
**Pull request by poettering**
A comprehensive fix for the credential part of #27260 and various other improvements.

Fixes #27260

### CODE REVIEW

File: src/shared/creds-util.c [hunk 1]
```diff
@@ -125,6 +125,80 @@ int read_credential(const char *name, void **ret, size_t *ret_size) {
                         (char**) ret, ret_size);
 }
 
+int read_credential_with_decryption(const char *name, void **ret, size_t *ret_size) {
```
Comment (daandemeyer) [inline src/shared/creds-util.c:128]: This sucks for generators outside of systemd's codebase since they won't have access to this function. Can't we set up a decrypted credentials directory for generators as well?

File: man/systemd.exec.xml [hunk 2]
```diff
@@ -3274,18 +3274,19 @@ StandardInputData=V2XigLJyZSBubyBzdHJhbmdlcnMgdG8gbG92ZQpZb3Uga25vdyB0aGUgcnVsZX
...
         use the <command>qemu</command> <literal>fw_cfg</literal> node
-        <literal>opt/io.systemd.credentials/</literal>. Example <command>qemu</command> switch: <literal>-fw_cfg
-        name=opt/io.systemd.credentials/mycred,string=supersecret</literal>. They may also be specified on
-        the kernel command line using the <literal>systemd.set_credential=</literal> switch (see
-        <citerefentry><refentrytitle>systemd</refentrytitle><manvolnum>1</manvolnum></citerefentry>) and from
-        the UEFI firmware environment via
-        <citerefentry><refentrytitle>systemd-stub</refentrytitle><manvolnum>7</manvolnum></citerefentry>.</para>
+        <literal>opt/io.systemd.credentials/</literal>. Example <command>qemu</command> switch:
+        <literal>-fw_cfg name=opt/io.systemd.credentials/mycred,string=supersecret</literal>. They may also
+        be specified on the kernel command line using the <literal>systemd.set_credential=</literal> switch
```
Comment (bluca) [inline man/systemd.exec.xml]: I would rephrase to put the cmdline last, and de-emphasize it somehow (saying that it's the least preferred or something), as it would be good to get people to stop using the kernel cmdline to configure userspace, and use the other mechanisms instead

File: src/core/main.c [hunk 3]
```diff
@@ -2224,10 +2224,13 @@ static int initialize_runtime(
                                 return r;
                         }
 
+                        /* Pull credentials from various sources into a common credential directory */
```
Comment (bluca) [inline src/core/main.c]: Maybe add a comment that this is done first as it can set the machine-id, as it might not be obvious during future refactors

File: test/test-functions [hunk 4]
```diff
@@ -577,7 +584,32 @@ run_qemu() {
     fi
 
     if [[ -n "$INITRD" ]]; then
-        qemu_options+=(-initrd "$INITRD")
+        if [[ -n "$INITRD_EXTRA" ]]; then
+            # An addition initrd has been specified, let's combine it with the main one.
+            local t=/var/tmp/initrd.combined."$RANDOM"
```
Comment (bluca) [inline test/test-functions]: `local t="$(mktemp --tmpdir=$WORKDIR -d -t initrd.combined.XXXXXX)"`

File: docs/CREDENTIALS.md [hunk 5]
```diff
@@ -288,6 +320,13 @@ services where they are ultimately consumed.
    initrds, as userspace can place credentials next to these EFI kernels, and
    be sure they can be accessed securely from initrd context.
 
+5. Credentials may be passed from the initrd to the host during the initrd →
```
Comment (bluca) [inline docs/CREDENTIALS.md]: as earlier I'd swap the cmdline and this one, and say that the cmdline should be a last resort or so

File: man/systemd.generator.xml [hunk 6]
```diff
@@ -190,6 +190,21 @@
         <varname>ConditionArchitecture=</varname> in
         <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry>.</para></listitem>
       </varlistentry>
+
+
```
Comment (bluca) [inline man/systemd.generator.xml]: extra newline

File: man/systemd.xml [hunk 7]
```diff
@@ -1051,7 +1051,40 @@
...
+
+    <itemizedlist>
+      <listitem><para>When the service manager first initializes it will read system credentials from SMBIOS
+      Type 11 vendor strings
+      <varname>io.systemd.credential:<replaceable>name</replaceable>=<replaceable>value</replaceable></varname>,
+      and
+      <varname>io.systemd.credential.binary:<replaceable>name</replaceable>=<replaceable>value</replaceable></varname>.</para></listitem>
+
+      <listitem><para>At the same time it will import credentials from QEMU <literal>fw_cfg</literal>. (Note
+      that the SMBIOS mechanism is generally preferred, because faster and generic.)</para></listitem>
```
Comment (bluca) [inline man/systemd.xml]: "because it is"

**Review by bluca — APPROVED** [review]
needs rebase

File: src/shared/creds-util.c (hunk 1 above)
Comment (poettering) [reply src/shared/creds-util.c:128]: that would mean decrypting *all* passed credentials unconditionally. That's not only problematic because decrypting all this sensitive data of which most might not be consumed at all by the generators means we reveal secrets to way more code than necessary. more problematic is that creds are bound to TPM, and come with PCR policies by default, which can be bound to runtime phases. we should not try to decrypt those, since that's slow, and will quite possibly fail (simply because the cred in question was itended to be unlocked at a different time), and increase "dictionary lockout" counters in the tpm and so on (which will fuck up the object entirely).

generators currently run with almost full system privileges, they are conceptually more part of pid 1 than part of the OS payload. They run in the /init.scope/ cgroup and such. Hence I think it's somewhat OK to have a slightly higher level of requirement on this. But yeah, it's not ideal.

I think for now, people should just shell out to "systemd-creds --system cat" and it will do the right thing... One of those days we should maybe offer some form of IPC API for this in PID 1 or so dunno.

File: test/test-functions (hunk 4 above)
Comment (poettering) [reply test/test-functions]: "-d" is not right becuase this is about a file, not a dir.

And mktemp creates the file, which is really not interesting here, we want to create the file ourselves.

using the $WORKDIR migh tmake sense though

Comment (bluca) [reply test/test-functions]: mktemp can create files too, just omit -d?

**Comment by bluca** [comment]
wrong PR for the last commit?

**Comment by poettering** [comment]
> wrong PR for the last commit?

yeah, i fucked that up.

**Comment by bluca** [comment]
```
21:36:39 [   19.518882] testsuite-54.sh[1124]: ++ systemd-creds cat --system myinitrdcred
21:36:39 [   19.546276] testsuite-54.sh[422]: + test '' = guatemala
21:36:39 [[0;1;31mFAILED[0m] Failed to start [0;1;39mtestsuite-54.service[0m.
```

File: test/test-functions (hunk 4 above)
Comment (poettering) [reply test/test-functions]: but why use mktemp? it creates a file for us, which we cannot use, we want to create the file ourselves. what's the benefit of using mktemp if it does the one thing we really don't want it to do...

**Comment by poettering** [comment]
Force pushed a new version that hopefully makes CI pass this time.

To simplify review, the only commits that changed since the first green-labelling are "test: add test for initrd credentials" and the new one "import-creds: add suppotr for binary credentials specified on the kernel cmdline" (i.e. the newest one in the PR)

**Review by bluca — APPROVED** [review]

**Comment by bluca** [comment]
computer says no:

```
12:31:23 [   18.901467] testsuite-54.sh[1115]: + cmp /etc/os-release
12:31:23 [   18.916030] testsuite-54.sh[1114]: Running as unit: test-54-exec-start.service
12:31:23          Starting [0;1;39mtest-54-exec-start.service[0m...
12:31:23 [   19.102845] testsuite-54.sh[1118]: Failed to open credentials directory '/run/credentials/test-54-exec-start.service': Permission denied
12:31:23 [   19.103143] testsuite-54.sh[1118]: Failed to open credentials directory: Permission denied
12:31:23 [[0;1;31mFAILED[0m] Failed to start [0;1;39mtest-54-exec-start.service[0m.
12:31:23 See 'systemctl status test-54-exec-start.service' for details.
12:31:23 [   19.215550] testsuite-54.sh[1114]: Finished with result: exit-code
12:31:23 [   19.215789] testsuite-54.sh[1114]: Service runtime: 135ms
12:31:23 [   19.215860] testsuite-54.sh[1114]: CPU time consumed: 122ms
12:31:23 [   19.226630] testsuite-54.sh[1115]: cmp: EOF on - which is empty
12:31:23 [[0;1;31mFAILED[0m] Failed to start [0;1;39mtestsuite-54.service[0m.
```

**Comment by poettering** [comment]
Wohoo it passed#

**Comment by poettering** [comment]
Only new change since last push is the new fix ACL commit early in the series. Rest unchanged.

**Comment by bluca** [comment]
This in centos8 looks new, any idea @mrc0mmand :

```
22:44:10 FAIL: test_interleaved (__main__.ExecutionResumeTest)
22:44:10 ----------------------------------------------------------------------
22:44:10 Traceback (most recent call last):
22:44:10   File "./test/test-exec-deserialization.py", line 166, in test_interleaved
22:44:10     self.check_output(expected_output)
22:44:10   File "./test/test-exec-deserialization.py", line 112, in check_output
22:44:10     self.assertEqual(output, expected_output)
22:44:10 AssertionError: 'foo\n' != 'foo\nbar\n'
22:44:10   foo
22:44:10 + bar
```

**Comment by mrc0mmand** [comment]
> This in centos8 looks new, any idea @mrc0mmand :
> 
> ```
> 22:44:10 FAIL: test_interleaved (__main__.ExecutionResumeTest)
> 22:44:10 ----------------------------------------------------------------------
> 22:44:10 Traceback (most recent call last):
> 22:44:10   File "./test/test-exec-deserialization.py", line 166, in test_interleaved
> 22:44:10     self.check_output(expected_output)
> 22:44:10   File "./test/test-exec-deserialization.py", line 112, in check_output
> 22:44:10     self.assertEqual(output, expected_output)
> 22:44:10 AssertionError: 'foo\n' != 'foo\nbar\n'
> 22:44:10   foo
> 22:44:10 + bar
> ```

I've seen this a couple of times already in other PRs, feel free to ignore it here.

**Review by bluca — APPROVED** [review]
